Source/WTF/wtf/text/StringCommon.h

```diff
@@ -843,6 +848,218 @@ inline bool equalIgnoringASCIICase(ASCIILiteral a, ASCIILiteral b)
+        destination[i] = static_cast<uint8_t>(source[i]);
+#endif
+}
+
+inline void copyElements(uint16_t* destination, const uint32_t* source, unsigned length)
+{
+    const auto* end = destination + length;
+#if CPU(ARM64)
+    const uintptr_t memoryAccessSize = 32 / sizeof(uint32_t);
+    if (length >= memoryAccessSize) {
+        const uintptr_t memoryAccessMask = memoryAccessSize - 1;
+        const uintptr_t lengthLeft = end - destination;
+        const auto* const simdEnd = destination + (lengthLeft & ~memoryAccessMask);
+        // Use ld2 to load lower 16bit of 8 uint32_t.
+        do {
+            asm("ld2   { v0.8H, v1.8H }, [%[SOURCE]], #32\n\t"
+                "st1   { v0.8H }, [%[DESTINATION]], #16\n\t"
+                : [SOURCE]"+r" (source), [DESTINATION]"+r" (destination)
+                :
+                : "memory", "v0", "v1");
+        } while (destination != simdEnd);
+    }
+#endif
+    while (destination != end)
+        *destination++ = *source++;
+}
+
+inline void copyElements(uint32_t* destination, const uint64_t* source, unsigned length)
+{
+    const auto* end = destination + length;
+#if CPU(ARM64)
+    const uintptr_t memoryAccessSize = 32 / sizeof(uint64_t);
+    if (length >= memoryAccessSize) {
+        const uintptr_t memoryAccessMask = memoryAccessSize - 1;
+        const uintptr_t lengthLeft = end - destination;
+        const auto* const simdEnd = destination + (lengthLeft & ~memoryAccessMask);
+        // Use ld2 to load lower 32bit of 4 uint64_t.
+        do {
+            asm("ld2   { v0.4S, v1.4S }, [%[SOURCE]], #32\n\t"
+                "st1   { v0.4S }, [%[DESTINATION]], #16\n\t"
+                : [SOURCE]"+r" (source), [DESTINATION]"+r" (destination)
+                :
+                : "memory", "v0", "v1");
+        } while (destination != simdEnd);
+    }
+#endif
+    while (destination != end)
+        *destination++ = *source++;
+}
+
+inline void copyElements(uint16_t* destination, const uint64_t* source, unsigned length)
+{
+    const auto* end = destination + length;
+#if CPU(ARM64)
+    const uintptr_t memoryAccessSize = 64 / sizeof(uint64_t);
+    if (length >= memoryAccessSize) {
+        const uintptr_t memoryAccessMask = memoryAccessSize - 1;
+        const uintptr_t lengthLeft = end - destination;
+        const auto* const simdEnd = destination + (lengthLeft & ~memoryAccessMask);
+        // Use ld4 to load lower 16bit of 8 uint64_t.
+        do {
+            asm("ld4   { v0.8H, v1.8H, v2.8H, v3.8H }, [%[SOURCE]], #64\n\t"
+                "st1   { v0.8H }, [%[DESTINATION]], #16\n\t"
+                : [SOURCE]"+r" (source), [DESTINATION]"+r" (destination)
+                :
+                : "memory", "v0", "v1", "v2", "v3");
+        } while (destination != simdEnd);
+    }
+#endif
+    while (destination != end)
+        *destination++ = *source++;
+}
+
+inline void copyElements(uint8_t* destination, const uint64_t* source, unsigned length)
```

Comment:
maybe `restrict` could be used and the non-SIMD while loops would get auto-vectorized?

Reply:
I also wonder if the assembly is different/better if size_t is used instead of unsigned

Reply:
`size_t` does not matter since this will be inlined.

Reply:
For restrict, I think it does not affect on the result since SIMD asm path will be used. But anyway, it would be good for non SIMD architectures.

Reply:
👍

Reply:
Actually, since now copyElements can be used for non string, size_t would be better as a API (String is always unsigned length, but others may not be true).